File: adt2unit/put_chunk.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// pattern: 0..max_patterns-1
// line: 0..255
// channel: 1..20
void put_chunk (uint8_t pattern, uint8_t line, uint8_t channel, const tCHUNK *chunk)
{
  if (pattern < max_patterns)
  {
    (*pattdata)[pattern / 8][pattern % 8][channel - 1][line] = *chunk;
#if !ADT2PLAY
    module_archived = false;
#endif // !ADT2PLAY
  }
  else
    limit_exceeded = true;
}

File: adt2unit/i386/put_chunk.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure put_chunk(pattern,line,channel: Byte; chunk: tCHUNK);
begin
  asm
        lea     esi,[chunk]
        mov     edi,[pattdata]
        mov     al,pattern
        inc     al
        cmp     al,max_patterns
        jbe     @@1
        mov     limit_exceeded,TRUE
        jmp     @@2
@@1:    xor     eax,eax
        mov     al,line
        mov     ebx,CHUNK_SIZE
        mul     ebx
        mov     ecx,eax
        xor     eax,eax
        mov     al,channel
        dec     eax
        mov     ebx,256*CHUNK_SIZE
        mul     ebx
        add     ecx,eax
        xor     eax,eax
        mov     al,pattern
        mov     ebx,8
        div     ebx
        push    eax
        mov     eax,edx
        mov     ebx,20*256*CHUNK_SIZE
        mul     ebx
        add     ecx,eax
        pop     eax
        mov     ebx,8*20*256*CHUNK_SIZE
        mul     ebx
        add     ecx,eax
        add     edi,ecx
        mov     ecx,CHUNK_SIZE
        rep     movsb
{$IFNDEF ADT2PLAY}
        mov     module_archived,FALSE
{$ENDIF} // NOT DEFINED(ADT2PLAY)
@@2:
  end;
end;
